src/github.com/getlantern/ops/ops.go

```diff
@@ -0,0 +1,150 @@
+// Package ops provides a facility for tracking the processing of operations,
+// including contextual metadata about the operation and their final success or
+// failure. An op is assumed to have succeeded if by the time of calling Exit()
+// no errors have been reported. The final status can be reported to a metrics
+// facility.
+package ops
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/getlantern/context"
+)
+
+var (
+	cm             = context.NewManager()
+	reporters      []Reporter
+	reportersMutex sync.RWMutex
+)
+
+// Reporter is a function that reports the success or failure of an Op. If
+// failure is nil, the Op can be considered successful.
+type Reporter func(failure error, ctx map[string]interface{})
+
+// Op represents an operation that's being performed. It mimics the API of
+// context.Context.
+type Op interface {
+	// Begin marks the beginning of an Op under this Op.
+	Begin(name string) Op
+
+	// Go starts the given function on a new goroutine.
+	Go(fn func())
+
+	// End marks the end of this op, at which point the Op will report its success
+	// or failure to all registered Reporters.
+	End()
+
+	// Set puts a key->value pair into the current Op's context.
+	Set(key string, value interface{}) Op
+
+	// SetDynamic puts a key->value pair into the current Op's context, where the
+	// value is generated by a function that gets evaluated at every Read.
+	SetDynamic(key string, valueFN func() interface{}) Op
+
+	// FailIf marks this Op as failed if the given err is not nil. If FailIf is
+	// called multiple times, the latest error will be reported as the failure.
+	// Returns the original error for convenient chaining.
+	FailIf(err error) error
```

Comment:
`FailIf` and `Errorf` may create confusing what is the best way to write error handling code, as there are multiple ways for similar purpose (saw some of them in this PR):
* `log.Errorf(...)`
* `log.Error(errors.New(...))`
* `log.Error(op.FailIf(errors.New(...)))`
* `log.Error(op.Errorf(...)))`

[Another example](https://github.com/getlantern/lantern/pull/4363/files#diff-c27ca30fd23b2036205dc308988a8b0dR123):
```
if err != nil {
  return nil, op.FailIf(err)
}
```
It's awkward to call `FailIf` when `err` is not nil for sure.

As simple wrappers, `FailIf` and `Errorf` creates more problems than solved. I vote to remove `op.Errorf` and rename `op.FailIf()` to `op.Fail()` without returning anything. So someone needs to do multiple things should call each individually, see below: 
```
n, err := doSomething()
if err != nil {
  op.Fail(err)
  log.Error(err)
}
return n, err
```

By provide only one way to do one thing, we make the code more understandable and encourage caller to think a bit on questions like "why return the same error after marking the `op` as `Fail` and logging the error?" or "is it the right place to call `op.Fail`?"